api/dataprocessor_test.go

```diff
@@ -1095,6 +1096,227 @@ func TestMergeSeries(t *testing.T) {
+	}
+}
+
+func TestRequestContextWithConsolidator(t *testing.T) {
+	metric := "metric1"
+	archInterval := uint32(10)
+	from := uint32(44)
+	to := uint32(88)
+	req := reqRaw(metric, from, to, 100, 10, consolidation.Sum)
+	req.ArchInterval = archInterval
+	ctx := newRequestContext(&req, consolidation.Sum)
+
+	expectFrom := from
+	if ctx.From != expectFrom {
+		t.Errorf("requestContext.From is not original value as expected, expected/got %d/%d", expectFrom, ctx.From)
+	}
+
+	expectTo := to
+	if ctx.To != expectTo {
+		t.Errorf("requestContext.To is not original value as expected, expected/got %d/%d", expectTo, ctx.To)
+	}
+
+	expectedAggKey := fmt.Sprintf("%s_%s_%d", metric, "sum", archInterval)
+	if ctx.AggKey != expectedAggKey {
+		t.Errorf("requestContext.AggKey is not as expected, expected/got \"%s\"/\"%s\"", expectedAggKey, ctx.AggKey)
+	}
+}
+
+// the essential idea of this test is that we want to put some chunks into the
+// cache and some into the backend store.
+// then we call getSeriesCachedStore and check if it has stitched the pieces
+// from the different sources together correctly.
+//
+// cache:           |-----|-----|-----|
+// store:                             |-----|
+// query:              |---------------|
+// expected result: |-----|-----|-----|-----|
+func TestGetSeriesCachedStore(t *testing.T) {
+	chunkSpan := uint32(600)
+	numChunks := uint32(10)
+	stats, _ := helper.New(false, "", "standard", "metrictank", "")
+	srv, _ := NewServer(stats)
+	store := mdata.NewMockStore()
+	srv.BindBackendStore(store)
+	metrics := mdata.NewAggMetrics(store, chunkSpan, numChunks, 0, 0, 0, 0, []mdata.AggSetting{})
+	mdata.InitMetrics(stats)
+	srv.BindMemoryStore(metrics)
+	defer metrics.Stop()
+	c := cache.NewCCache()
+	srv.BindCache(c)
+	metricKey := "metric1"
+
+	// generating 4 chunks, 3 will go into the cache and 1 into the backend store.
+	// this simulates a scenario where the latest chunk has been persisted but not
+	// queried (and cached) yet.
+	chunks := make([]*chunk.Chunk, 0)
+	for i := chunkSpan; i < chunkSpan*5; i++ {
+		if i%chunkSpan == 0 {
+			chunks = append(chunks, chunk.New(i))
+		}
+		chunks[len(chunks)-1].Push(i, float64(i*2))
+	}
+
+	prevts := uint32(0)
+	// adding 3 itgens to cache
+	var i int
+	for i = 0; i < 3; i++ {
+		chunks[i].Series.Finish()
+		itgen := chunk.NewBareIterGen(chunks[i].Series.Bytes(), chunks[i].Series.T0, chunkSpan)
+		srv.Cache.Add(metricKey, prevts, *itgen)
+		prevts = itgen.Ts()
+	}
+
+	// preparing the next itgen as store mock result (to simulate one that has just been added and not cached yet)
+	chunks[i].Series.Finish()
+	itgens := []chunk.IterGen{*chunk.NewBareIterGen(chunks[i].Series.Bytes(), chunks[i].Series.T0, chunkSpan)}
+
+	// save some electrons by skipping steps that are no edge cases
+	steps := chunkSpan / 10
+
+	// we want to have:
+	// - queries that only query chunks in the cache
+	// - queries that only query chunks in cassandra
```

Comment:
I don't we succeed in triggering this scenario because to only query chunks in cassandra, we should have a result that is only in the last chunk, which starts at 4*chunkSpan, but the for loop will never go that far.

I suggest we change the for loops to this, which cover a bit more ground:
```
	for from := chunkSpan; from < chunkSpan*5-1; from += steps {
		for to := from + 1; to <= chunkSpan*5; to += steps {
```